### ISSUE TITLE
`Failed to set up event listener: Unexpected token W in JSON at position 0` error

### ISSUE THREAD

**Issue by bwateratmsft** [microsoft/vscode-docker#3725]
Originally from https://github.com/microsoft/vscode-docker/issues/3724#issuecomment-1330889789

/cc @brainz80

**Comment by bwateratmsft** [comment]
@brainz80 can you share the output from this command?

```sh
docker events --filter type=container --filter event=create --format "{'Type':{{json .Type}},'Action':{{json .Action}},'Actor':{{json .Actor}},'Time':{{json .Time}},'Raw':{{json .}}}"
```

In a separate terminal, try starting a container, e.g. `docker run hello-world`, to cause some events to be fired.

**Comment by scottross8** [comment]
I have a similar problem except that my message is "Failed to connect. Is Docker running? Error: Invalid container JSON"   The filter in the `docker events` command above generates an invalid JSON message (wrong quotes.)   In my case I am using podman (named docker but it is podman).  Podman doesn't seem to pass the Actor nor Action fields.  I changed the format to only show the raw message:

```
$ docker events --filter type=container --filter event=create --format '{"Raw":{{json .}}}' | jq
{
  "Raw": {
    "ID": "557b660937639152190968ff7f107eafeeee3d5a7bc6a2322b6fc8cef70e954f",
    "Image": "docker.io/library/busybox:latest",
    "Name": "frosty_swartz",
    "Status": "create",
    "Time": "2022-11-29T11:15:58.935745953-07:00",
    "Type": "container",
    "Attributes": null
  }
}
```
If this is enough of a different problem, and it would help you in tracking the issues, I will create a new issue for you.

**Comment by bwateratmsft** [comment]
Podman's JSON output is different enough that it fails our parsing. We are looking into the right way to address that; tracking it currently in #3241. Unfortunately it's just not possible for Podman to work with the explorer view right now in 1.23.0, but you can roll back to 1.22.2 with the palette command "Extensions: Install Specific Version of Extension...".

**Comment by brainz80** [comment]
@bwateratmsft Just ran 

```cmd
docker events --filter type=container --filter event=create --format "{'Type':{{json .Type}},'Action':{{json .Action}},'Actor':{{json .Actor}},'Time':{{json .Time}},'Raw':{{json .}}}"
```

together with `docker run hello-world` this is what I got:

```json
{'Type':"container",'Action':"create",'Actor':{"ID":"7b83c178504c5242127a0b873c92be8096e2d00cd63d3d40ac5d6f7fc30f9565","Attributes":{"image":"hello-world","name":"flamboyant_hugle"}},'Time':1669794567,'Raw':{"status":"create","id":"7b83c178504c5242127a0b873c92be8096e2d00cd63d3d40ac5d6f7fc30f9565","from":"hello-world","Type":"container","Action":"create","Actor":{"ID":"7b83c178504c5242127a0b873c92be8096e2d00cd63d3d40ac5d6f7fc30f9565","Attributes":{"image":"hello-world","name":"flamboyant_hugle"}},"scope":"local","time":1669794567,"timeNano":1669794567962070900}}
```

Shouldn't the keys the surrounded with `""` and not `''` in valid JSON?

**Comment by bwateratmsft** [comment]
The above query is a simplified version of what the extension uses. You're correct, valid JSON should use `"` for keys instead of `'`. I'm not sure how we missed this, though the JSON parser we're using (just simple `JSON.parse()`) doesn't seem to mind this.

My suspicion is that some warning or error message was printed to stdout when this command was run, which the parser tried to parse and of course failed. We use this `docker events` feature to cause the relevant views in the explorer to refresh; we also refresh once a minute and on certain configuration changes. There's also the manual refresh ability too. So, we expect even with this error that the explorer view will work, it just won't auto-refresh (except once a minute or on manual refresh).

If you restart VSCode, does this error occur again and show up in the output log, when you open the explorer view?

**Comment by brainz80** [comment]
@bwateratmsft  I restarted my entire computer. The problem persists :(

**Comment by brainz80** [comment]
I believe I found the problem.

I took a look into VSCode:s Process Explorer. Found this:

```cmd
powershell.exe -c "docker events --since 1669900632 --until 1669900932 --filter 'type=container' --filter 'type=image' --filter 'type=network' --filter 'type=volume' --filter 'event=create' --filter 'event=destroy' --filter 'event=die' --filter 'event=kill' --filter 'event=pause' --filter 'event=rename' --filter 'event=restart' --filter 'event=start' --filter 'event=stop' --filter 'event=unpause' --filter 'event=update' --filter 'event=delete' --filter 'event=import' --filter 'event=load' --filter 'event=pull' --filter 'event=save' --filter 'event=tag' --filter 'event=untag' --filter 'event=remove' --filter 'event=prune' --format '{\\\"Type\\\":{{json .Type}},\\\"Action\\\":{{json .Action}},\\\"Actor\\\":{{json .Actor}},\\\"Time\\\":{{json .Time}},\\\"Raw\\\":{{json .}}}'"
```

Running ```powershell.exe``` tries to load my profile which outputs errors - because of uninstalled modules. I'm using pwsh.exe (Powershell 7). Should this command be run with `-NoProfile` flag to prevent this from happening?

**Comment by bwateratmsft** [comment]
Good find! @danegsta is there any reason we can't/shouldn't use the `-NoProfile` flag?

**Comment by brainz80** [comment]
@bwateratmsft Could `powershell.exe` be replaced with eg. `powershell.cmd`?
Which could look like:

```batch
@echo off
powershell.exe -NoProfile %*
```

This file would have to added to the `vsix` -bundle

**Comment by bwateratmsft** [comment]
We've been looking at whether we even need to use a shell at all. I was under the (mistaken) impression that Node would not look at `$PATH`, so a shell would have to be used to launch `docker` processes, but that doesn't seem to be the case. It would greatly simplify the work we have to do of dealing with argument quoting/escaping/etc. if we didn't have to use a shell. It might also be very slightly faster-performing.

**Comment by bwateratmsft** [comment]
I've made a change so that for most background-executed `docker` processes, we execute _without_ a shell. There are still a handful of things that will still execute _with_ a shell, but almost all of them relate to the `docker-build` / `docker-run` / `docker-compose` tasks, and debugging. As before, we'll use PowerShell on Windows as the shell to execute with, but, most of those cases don't involve actually parsing (or doing anything with) the output, so profiles emitting some extra stuff shouldn't matter.

**Comment by bwateratmsft** [comment]
A fix for this has been released in Docker extension version 1.23.2.
